[Feature] support more attributes for comparing StarRocks tables by using Alembic

jaogoy opened the pull request:
## Why I'm doing:
Users want to manage their database schema with SQLAlchemy+Alembic for schema migration.
Starrocks-python-client currently only support very limited functions.
So, we need to enrich the sqlalchemy dialect features of StarRocks, and support comparing table attributes by using Alembic.

## What I'm doing:

Fixes #issue

### Summary

This PR introduces comprehensive support for Alembic migrations with StarRocks, enabling developers to manage schema changes for tables, including complex nested types and table properties. It significantly enhances the SQLAlchemy dialect to recognize more StarRocks-specific table attributes, particularly for partitioning and distribution.

While significant progress has been made on `VIEW` and `MATERIALIZED VIEW` support, these features are not yet complete. A refactoring is underway to align their usage with the `Table` construct for a more consistent API, and so they are not ready for use in this PR.

### Key Features

*   **Alembic Integration for Tables:**
    *   Adds robust comparison logic for StarRocks tables within
    Alembic's `autogenerate` workflow.
    *   Detects changes in table properties, such as `distributed_by`,
    `partition_by`, `order_by`, and other table options. The comparison
    logic now correctly handles immutable properties like `engine`,
    preventing erroneous diffs.
    *   Compares column definitions, including complex and nested types.

*   **Enhanced SQLAlchemy Dialect:**
    *   The dialect's reflection capabilities have been enhanced,
    particularly for `partition` and `distribution` properties, to provide
    more accurate table metadata.
    *   Added support for defining StarRocks-specific table properties
    directly from SQLAlchemy's `Table` object.
    *   Improved type compilation for both simple and complex data
    types.

*   **Complex Nested Data Type Support:**
    *   Implemented comparison for changes within complex column types
    like `ARRAY`, `MAP`, and `STRUCT`, including **nested** structures. For
    example, it can detect changes in an `ARRAY` of `STRUCT`s.

*   **Extensive Test Coverage:**
    *   Added a large suite of unit, integration, and system tests to
    verify the new functionality.
    *   Tests cover table property comparison, column type changes, and
    the end-to-end Alembic `autogenerate` process.

*   **Initial Support for Views and Materialized Views (In-progress):**
    *   Substantial implementation for the comparison of `VIEW`s and
    `MATERIALIZED VIEW`s has been completed.
    *   **Note:** This functionality is incomplete as a refactoring is
    required to align the API with the standard `Table` interface. It is not
    ready for use at this stage.

### Detailed Changes

*   **New `starrocks.alembic` module:** This module contains the core
logic for comparing and rendering StarRocks-specific schema objects.
    *   `compare.py`: Implements the comparison functions for tables and
    columns.
    *   `ops.py`: Defines custom Alembic operation objects for
    StarRocks.
    *   `render.py`: Provides logic to generate the Python migration
    code for Alembic.
    *   `starrocks.py`: The entrypoint that registers the StarRocks
    implementation with Alembic.
*   **Updates to `starrocks.dialect` and `starrocks.reflection`:** These
files were updated to handle the reflection of new table properties and
complex types.
*   **New Documentation:** Added extensive documentation on how to use
the new Alembic features and define tables with StarRocks properties.
*   **New Tests:** A significant number of tests were added to ensure
the quality of the new features.

### How to Use

For detailed instructions on using the new Alembic features for tables, please refer to the documentation: `docs/usage_guide/alembic.md`.


## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Add full Alembic integration to the StarRocks SQLAlchemy dialect with table diff/render/ops (incl. complex types), initial view/MV support, reflection/DDL enhancements, docs, and extensive tests.
> 
> - **Alembic Integration**:
>   - Add `starrocks.alembic` (compare, ops, render, impl) for autogenerate diffs and migration rendering.
>   - Support ALTER ops: `distribution`, `order by`, `properties`; block unsupported: `engine`, `key`, `partition` (with clear errors).
> - **Complex Types & Rendering**:
>   - Compare/render nested `ARRAY`/`MAP`/`STRUCT`; special equivalences (e.g., `BOOLEAN`↔`TINYINT(1)`, `STRING`↔`VARCHAR(65533)`).
> - **Views & Materialized Views**:
>   - Implement create/drop/alter (render) and autogenerate (initial support) for `VIEW`/`MATERIALIZED VIEW`.
> - **Dialect/Reflection Enhancements**:
>   - Improve reflection for `partition`, `distribution`, order/properties; add robust SQL normalization and parsing.
> - **Docs & Examples**:
>   - New guides: Alembic usage, SQLAlchemy tables, views/MVs; sample scripts.
> - **Tests & Packaging**:
>   - Extensive unit/integration/system tests; bump version to `1.3.0`, add build/test config and dev deps.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1c41ba4bee1b9c66586d2f22922641b6a7d262ec. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jaogoy committed:
[Feature] support more attributes for comparing StarRocks tables by using Alembic

This PR introduces comprehensive support for Alembic migrations with
StarRocks, enabling developers to manage schema changes for tables,
including complex nested types and table properties. It significantly
enhances the SQLAlchemy dialect to recognize more StarRocks-specific
table attributes, particularly for partitioning and distribution.

While significant progress has been made on `VIEW` and `MATERIALIZED
VIEW` support, these features are not yet complete. A refactoring is
underway to align their usage with the `Table` construct for a more
consistent API, and so they are not ready for use in this PR.

*   **Alembic Integration for Tables:**
*   Adds robust comparison logic for StarRocks tables within
Alembic's `autogenerate` workflow.
*   Detects changes in table properties, such as `distributed_by`,
`partition_by`, `order_by`, and other table options. The comparison
logic now correctly handles immutable properties like `engine`,
preventing erroneous diffs.
    *   Compares column definitions, including complex and nested types.

*   **Enhanced SQLAlchemy Dialect:**
*   The dialect's reflection capabilities have been enhanced,
particularly for `partition` and `distribution` properties, to provide
more accurate table metadata.
*   Added support for defining StarRocks-specific table properties
directly from SQLAlchemy's `Table` object.
*   Improved type compilation for both simple and complex data
types.

*   **Complex Nested Data Type Support:**
*   Implemented comparison for changes within complex column types
like `ARRAY`, `MAP`, and `STRUCT`, including **nested** structures. For
example, it can detect changes in an `ARRAY` of `STRUCT`s.

*   **Extensive Test Coverage:**
*   Added a large suite of unit, integration, and system tests to
verify the new functionality.
*   Tests cover table property comparison, column type changes, and
the end-to-end Alembic `autogenerate` process.

*   **Initial Support for Views and Materialized Views (In-progress):**
*   Substantial implementation for the comparison of `VIEW`s and
`MATERIALIZED VIEW`s has been completed.
*   **Note:** This functionality is incomplete as a refactoring is
required to align the API with the standard `Table` interface. It is not
ready for use at this stage.

*   **New `starrocks.alembic` module:** This module contains the core
logic for comparing and rendering StarRocks-specific schema objects.
*   `compare.py`: Implements the comparison functions for tables and
columns.
*   `ops.py`: Defines custom Alembic operation objects for
StarRocks.
*   `render.py`: Provides logic to generate the Python migration
code for Alembic.
*   `starrocks.py`: The entrypoint that registers the StarRocks
implementation with Alembic.
*   **Updates to `starrocks.dialect` and `starrocks.reflection`:** These
files were updated to handle the reflection of new table properties and
complex types.
*   **New Documentation:** Added extensive documentation on how to use
the new Alembic features and define tables with StarRocks properties.
*   **New Tests:** A significant number of tests were added to ensure
the quality of the new features.

For detailed instructions on using the new Alembic features for tables,
please refer to the documentation: `docs/usage_guide/alembic.md`.

Signed-off-by: jaogoy <jaogoy@gmail.com>

rad-pat commented:
This seems to be already released to PyPI since earlier and has broken one of our CI tests which checks creation of materialized views in Starrocks with options.

jaogoy commented:
> I think alembic should be an optional requirement, not everyone using the dialect should need to install it. It should be installable via an option if required e.g. `pip install starrocks[alembic]` or similar. This however means it should not be imported into dialect.py

@rad-pat Yes, it's a good idea to do that. But, this project is not big, it won't hurt a lot, we may do it in the future.

jaogoy commented on contrib/starrocks-python-client/README.md:
```diff
@@ -1,17 +1,50 @@
 
+#### Supported Python Versions
+
+Python >= 3.8, <= 3.12
```
Do you have any idea if we make a change that the minimal version should be 3.10?
That, I've simply tested sqlacodegen, which may need 3.10 to be compatible with starrocks-sqlalchemy.

jaogoy commented on contrib/starrocks-python-client/README.md:
```diff
@@ -20,87 +53,269 @@ To connect to StarRocks using SQLAlchemy, use a connection string (URL) followin
+
+        'starrocks_PRIMARY_KEY': 'id',
+        'starrocks_DISTRIBUTED_BY': 'HASH(id) BUCKETS 10',
+        'starrocks_PROPERTIES': {'replication_num': '1'}
```
Both are supported, but the prefix `starrocks_` needs to be lower cased, it the requirement from sqlalchemy.

jaogoy commented on contrib/starrocks-python-client/README.md:
```diff
@@ -20,87 +53,269 @@ To connect to StarRocks using SQLAlchemy, use a connection string (URL) followin
-### Unit tests
-To run tests for Starrocks SQLAlchamy dialect (this module), run:
+
+### Tests
```
what does 'use uv or poetry`  mean? Could you share with me more?

TiansuYu replied:
Use [uv](https://docs.astral.sh/uv/) or [poetry](https://github.com/python-poetry/poetry) as the package manager. You will have a much smoother dev experience than trying to venv, build and publish every time. Use UV if you value speed.

Astralidea commented on contrib/starrocks-python-client/starrocks/alembic/compare.py:
```diff
@@ -0,0 +1,1214 @@
+        # Update the alter_column_op with the new aggregate type. useless now
+        # "KEY", "SUM" for set, None for unsert
+        if alter_column_op is not None:
+            alter_column_op.kwargs[ColumnAggInfoKeyWithPrefix.AGG_TYPE] = meta_agg_type
```
```suggestion
            alter_column_op.kw[ColumnAggInfoKeyWithPrefix.AGG_TYPE] = meta_agg_type
```

Astralidea commented on contrib/starrocks-python-client/starrocks/alembic/compare.py:
```diff
@@ -0,0 +1,1214 @@
+                    properties=metadata_mv.properties,
+                ),
+            )
+        )
```
```suggestion
        upgrade_ops.ops.append(DropMaterializedViewOp(mv_name, schema=schema))
        upgrade_ops.ops.append(CreateMaterializedViewOp(
            metadata_mv.name,
            metadata_mv.definition,
            schema=schema,
            comment=metadata_mv.comment,
            partition_by=metadata_mv.partition_by,
            distributed_by=metadata_mv.distributed_by,
            order_by=metadata_mv.order_by,
            refresh_moment=metadata_mv.refresh_moment,
            refresh_type=metadata_mv.refresh_type,
            properties=metadata_mv.properties,
       ))
```

rad-pat commented on contrib/starrocks-python-client/starrocks/dialect.py:
```diff
@@ -207,109 +325,294 @@ def visit_create_table(self, create, **kw):
-        ):
-            colspec[1] = "BIGINT" # ToDo - remove this, find way to fix the test
+        # AUTO_INCREMENT or default value or computed column
+        if column.autoincrement is True:
```
column.autoincrement is not bool, previous determination is more appropriate.

rad-pat commented on contrib/starrocks-python-client/starrocks/dialect.py:
```diff
@@ -319,57 +622,427 @@ def get_column_specification(self, column, **kw):
+          an aggregate function (e.g., 'SUM', 'REPLACE') specified via the
+          `info={'starrocks_agg': '...'}` dictionary on a Column.
+        - **AUTO_INCREMENT**: Automatically renders `AUTO_INCREMENT` for columns
+          with `autoincrement=True`. It also ensures these columns are `BIGINT`
```
This is not true for alter column, it is raising not supported error

jaogoy replied:
About auto_increment, you can optimized after this PR merged.

rad-pat reviewed (commented):
A couple of comments on autoinc

rad-pat commented:
> > > > @jaogoy, I took a copy of your branch so I could attempt to overlay my changes. Initially when I correct and enable the SQLAlchemy test suite again, there are quite a number of failures with reflection.
> > > 
> > > 
> > > @rad-pat Yes, about the sqlalchemy test_suite.py, I need to fix them. And there mey be some changes about the table key and table comment, maybe other things as well.
> > > Or, it the same thing as #63967?
> > 
> > 
> > It is similar to #63967. I am trying to overlay my changes onto yours so the changes are not so big, I will create a branch with this and maybe then a PR to your repo for review? I had this very nearly complete Friday so I expect to complete today if I do not have other distractions.
> 
> @rad-pat Hi, Pat, I need you help. I'm not fimilar with SQLAlchemy' testing, so when I run the test_suite.py by using `pytest test/test_suite.py`, it fails with:
> 
> ```shell
> file ./starrocks-python-client/test/test_suite.py, line 197
>       def test_default_reflection(self, connection, metadata):
> E       fixture 'connection' not found
> ...
> ```
> 
> I've talked with gpt or gemini a lot of times, none of the solutions work. Do I need to configure the testing environment? or need to install the proper version of SQLAlchemy or some other testing packages? I'll be very appreciated to receive your solution. I've taken a day on it, but no good result.

@jaogoy if you look at changes in commit https://github.com/StarRocks/starrocks/commit/ab414e97c3f80c3ae06e4034b6623058d4712c91, this should get the SQLAlchemy tests running. Specifically changes in the test/ folder, provision.py, reflection.py, setup.cfg
Hopefully that will get you going. That commit should be able to cherry-pick to this PR branch and SQLAlchemy tests should run.

jaogoy committed:
some small bugfix, and change from py38 to py310

1. add default replication_num=1 for small test env.
2. bugfix in test_suite.py
3. fix some problems of README

jaogoy committed:
modify the README.md

Signed-off-by: jaogoy <lijiao@starrocks.com>

jaogoy commented:
@rad-pat @Astralidea @TiansuYu 
I've done some modification, you can have a review, mainly in https://github.com/StarRocks/starrocks/pull/64161/commits/6462a8f32d70c999bc2e3849b9a84127a38f166c
Currently, I only pass all the newly added tests and the old test_suite.py, all the SQLAlchemy's tests haven't passed yet. I'm not fimilar with it, we can try to pass them in the following days with your help.
Then, we can make quick and small steps forward.

jaogoy commented:
> @jaogoy the changes look fine to me. In order to get SQLAlchemy tests working, some things need to be changed. Firstly, in the requirements.py, tests cannot be imported here in order to override them, this must be done in the test_suite.py. Creating a provision.py that creates the connections that SQLAlchemy is a useful addition, especially when running with pytest-xdist with 4 workers to make the tests run quickly.
> 
> This commit should help with tests [main...PlaidCloud:starrocks:main-fix-tests](https://github.com/StarRocks/starrocks/compare/main...PlaidCloud:starrocks:main-fix-tests)
> 
> As I remember, comments are an issue, as is auto_increment, restriction on starrocks_XXXX that can be passed to dialect causes an issue. There are others too, but all are fixed in this commit, though it also includes changes for adding FILES clause [PlaidCloud@ab414e9](https://github.com/PlaidCloud/starrocks/commit/ab414e97c3f80c3ae06e4034b6623058d4712c91)
> 
> Let me know if you want me to apply the fixes to your branch, but it may be good to understand them. I had to pull auto_increment from the SHOW CREATE TABLE. I found that one auto increment test failed because it incremented by 100000, but not sure why - it is disabled in test_suite.py

@rad-pat Can you spend some time to get the SQLAlchemy tests runnable?
I haven't  change something on `requirements.py` except some running problems with `import`.
You can directly add `-n4` anytime.
And, I want to split the test cases clearly, so I categorize them into `unit`, `integration`, `system`, but leave the old test cases in the `test_suite.py`. I'll try to use SQLAlchemy test style in the future, and transform my old test cases step by step.

`starrocks_COMMENT` is not good for Alembic comparision. When users create a `Table` in metadata, Alembic directly compare `comment` from `table.comment` rather from `table.dialect_options[dialect]["COMMENT"]`. But I did some compatibility work for **compiling** (but no idea how to do it for **comparison**, because the alembic's built-in comment comparison will run without and order control).

I want to merge it PR ASAP, then you can submit other PRs to do some bugfixs, features.
